### PR TITLE
[search] Fixed results with postcode ranking.

### PR DESCRIPTION
**Pull request by vng**

### CODE REVIEW

File: libs/search/ranker.cpp [hunk 1]
```diff
@@ -582,6 +582,23 @@ class RankerResultMaker
...
+      if (!preInfo.m_postcodeRange.Empty())
+      {
+        auto const postcode = ft.GetMetadata(feature::Metadata::FMD_POSTCODE);
+        if (!postcode.empty())
+        {
+          TokenSlice const slice(m_params, preInfo.m_postcodeRange);
+          NameScores pcScores;
+          UpdateNameScores(postcode, StringUtf8Multilang::kDefaultCode, slice, pcScores);
+
+          if (pcScores.m_nameScore > nameScore)
```
Comment (biodranik) [inline libs/search/ranker.cpp]: **Re: [P2] Preserve name-match ordering when postcode also matches**

For POIs (the main postcode search use case), this overwrite doesn't change behavior — `RankingInfo::GetNameScore()` (`ranking_info.cpp:390-397`) already promotes `FULL_PREFIX` → `FULL_MATCH` for POIs when only SUBPOI/COMPLEX_POI tokens are matched:

```cpp
// ranking_info.cpp:390-397 — existing code, not part of this PR
if (m_nameScore == NameScore::FULL_PREFIX)
{
  // query = "rewe", pois = ["REWE", "REWE City", "REWE to Go"]
  return NameScore::FULL_MATCH;
}
```

Since postcode tokens go into `m_postcodeRange` (not into `m_tokenRanges[]`), the early-return guard at line 384-386 (which checks `m_tokenRanges[TYPE_BUILDING+]`) doesn't fire, so the promotion applies to exactly these queries. Both "Nero" and "Nero City" already end up as FULL_MATCH before the postcode overwrite even runs.

For non-POI features (streets/buildings), the concern is technically valid but practically negligible: the FULL_PREFIX→FULL_MATCH delta is only **0.002** in the linear model (`kNameScore[]` at `ranking_info.cpp:52-58`), and two features with the same name prefix sharing a postcode is an extremely rare scenario.

A clarifying comment like `// Postcodes only promote, never demote the score` would help future readers understand the intentional asymmetry with the `std::min` used for dependent features (line 524).

File: libs/search/ranker.cpp [hunk 2]
```diff
@@ -582,6 +582,23 @@ class RankerResultMaker
...
+            nameScore = pcScores.m_nameScore;
+          matchedLength += pcScores.m_matchedLength;
```
Comment (biodranik) [inline libs/search/ranker.cpp]: **Re: [P2] Account for postcode misprints in the ranking error total**

Valid concern, though the practical scenario is narrower than described: the geocoder uses `UniStringDFA` (exact match, zero errors) for postcode retrieval — see `feature_offset_match.hpp:323-326`:

> *"Full match required even for prefix token. Reasons: 1. For postcode every symbol is important, partial matching can lead to wrong results."*

Features with typo-distance postcodes can only enter the result set via **geographic proximity** to the postcode point (`geocoder.cpp:1150`, `RetrieveGeometryFeatures`), not by fuzzy postcode matching. This makes the scenario (two same-name features in the same postcode area with 1-char-different postcodes) quite rare.

That said, the fix is trivial and safe:
```cpp
if (pcScores.m_errorsMade.IsValid())
    errorsMade += pcScores.m_errorsMade;
```
Worth adding for correctness even if the practical impact is small. The score delta for 1 postcode error is ~0.13 points (`kErrorsMade / numTokens`), which is non-negligible when it does occur.

File: libs/search/ranker.cpp [hunk 3]
```diff
@@ -582,6 +582,23 @@ class RankerResultMaker
 
       info.m_geoParts = preInfo.m_geoParts;
 
+      // Check if the feature's postcode matches the postcode tokens in the query.
+      // This distinguishes "Nero with postcode G4 9HS" from "Nero near postcode G4 9HS area".
+      if (!preInfo.m_postcodeRange.Empty())
+      {
+        auto const postcode = ft.GetMetadata(feature::Metadata::FMD_POSTCODE);
```
Comment (biodranik) [inline libs/search/ranker.cpp]: **Bug: use-after-free crash (SIGSEGV) in `ProcessorTest_TestPostcodes`**

`ft.GetMetadata(FMD_POSTCODE)` dereferences a dangling `m_loadInfo` pointer. `FeatureType` holds a **non-owning raw pointer** to `SharedLoadInfo` (`feature.hpp:261`), which lives inside the `FeaturesLoaderGuard` stored in `m_loader`. The crash sequence:

1. **Line 296:** `ft` is loaded from the "Wonderland" MWM. `ft.m_loadInfo` points into `m_loader`'s internals.
2. **Line 547:** `LoadFeature(preInfo.m_cityId)` loads the city from the **world MWM**. Since it's a different MWM, `m_loader` is replaced → the old loader is destroyed → `ft.m_loadInfo` becomes a **dangling pointer**.
3. **Line 589 (here):** `ft.GetMetadata(FMD_POSTCODE)` dereferences `m_loadInfo->m_metaDeserializer` which is now garbage → **SIGSEGV** at `MetadataDeserializer::GetMetaById(this=0x0)` (`metadata_serdes.cpp:50`).

This is a **pre-existing latent issue** — the loader replacement has always invalidated `ft.m_loadInfo` after city loading (line 547). It never crashed because existing code after that point only reads already-parsed/cached data from `ft` (name, center, types). Line 567 (`GetNameScores(ft, ...)`) also uses `ft` after city loading but only accesses the cached name, not metadata requiring `m_loadInfo`. This new `GetMetadata()` call is the first to need lazy deserialization via `m_loadInfo` after the loader swap.

**Suggested fix:** move this postcode block before the dependent-feature loading (before line 541). The postcode scoring is independent of street/city scores, and this avoids the dangling pointer entirely:

```cpp
auto matchedLength = scores.m_matchedLength;

// --- move postcode scoring here, right after initial name scoring ---
if (!preInfo.m_postcodeRange.Empty())
{
  auto const postcode = ft.GetMetadata(feature::Metadata::FMD_POSTCODE);
  if (!postcode.empty())
  {
    TokenSlice const slice(m_params, preInfo.m_postcodeRange);
    NameScores pcScores;
    UpdateNameScores(postcode, StringUtf8Multilang::kDefaultCode, slice, pcScores);

    if (pcScores.m_nameScore > nameScore)
      nameScore = pcScores.m_nameScore;
    matchedLength += pcScores.m_matchedLength;
  }
}

auto const updateScoreForFeature = [&](FeatureType & ft, Model::Type type) { ... };
```

Reproducer: `search_integration_tests --filter=ProcessorTest_TestPostcodes$` → exit code 139 (SIGSEGV).

File: libs/search/ranker.cpp (hunk 2 above)
Comment (vng) [reply libs/search/ranker.cpp]: - Added ```errorsMade += ```
- Removed ```nameScore = std::max(pcScores.m_nameScore, nameScore)```
- Left a comment, still not sure what is better.

File: libs/search/ranker.cpp (hunk 3 above)
Comment (vng) [reply libs/search/ranker.cpp]: Fixed

File: libs/search/ranker.cpp (hunk 1 above)
Comment (vng) [reply libs/search/ranker.cpp]: Removed nameScore = std::max(pcScores.m_nameScore, nameScore), still not sure.

**Review by biodranik — APPROVED** [review]

File: libs/search/ranker.cpp [hunk 4]
```diff
@@ -515,6 +515,35 @@ class RankerResultMaker
...
+      // This distinguishes "Nero with postcode G4 9HS" from "Nero near postcode G4 9HS area".
+      if (!preInfo.m_postcodeRange.Empty())
+      {
+        auto const & innerRange = preInfo.InnermostTokenRange();
+        bool const postcodeOverlapsInnerRange =
+            innerRange.Begin() < preInfo.m_postcodeRange.End() && preInfo.m_postcodeRange.Begin() < innerRange.End();
+
+        if (!postcodeOverlapsInnerRange)
+        {
+          auto const postcode = ft.GetMetadata(feature::Metadata::FMD_POSTCODE);
```
Comment (biodranik) [inline libs/search/ranker.cpp:528]: **Re: [P2] Postcode rescoring leaks into ancestor-matched results**

The main scenario the reviewer describes is handled correctly: when a building matches through its street/city's postcode but has no `FMD_POSTCODE` of its own, `ft.GetMetadata(FMD_POSTCODE)` returns empty → the `!postcode.empty()` guard on line 529 skips rescoring entirely — no boost, no penalty. The `VillagePostcodes` (`processor_test.cpp:2464`) and `StreetPostcodes` (`processor_test.cpp:2507`) tests confirm this path works.

Minor edge case: a leaf feature with a **close-but-different** postcode (e.g., "67391" vs query "67390") would get both a small `matchedLength` boost and an `errorsMade` penalty. The penalty dominates (`kErrorsMade=-0.4` vs `kMatchedFraction=0.19`), so it's a net penalty for a feature that was legitimately found via its ancestor. In practice this is very rare — buildings on the same street almost always share a postcode — and all existing tests pass.
